debian/mongoexport.1
.\" Man page generated from reStructuredText.
.
.TH "MONGOEXPORT" "1" "Jun 21, 2018" "4.0" "mongerdb-manual"
.SH NAME
mongerexport \- MongerDB Export Utility
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SS On this page
.INDENT 0.0
.IP \(bu 2
\fI\%Synopsis\fP
.IP \(bu 2
\fI\%Considerations\fP
.IP \(bu 2
\fI\%Required Access\fP
.IP \(bu 2
\fI\%Read Preference\fP
.IP \(bu 2
\fI\%Options\fP
.IP \(bu 2
\fI\%Use\fP
.UNINDENT
.INDENT 0.0
.INDENT 3.5
.IP "Mac OSX Sierra and Go 1.6 Incompatibility"
.sp
Users running on Mac OSX Sierra require the 3.2.10 or newer version
of  mongerexport\&.
.UNINDENT
.UNINDENT
.SH SYNOPSIS
.sp
\fI\%mongerexport\fP is a utility that produces a JSON or CSV export
of data stored in a MongerDB instance.
.sp
See the mongerimport document for more
information regarding the \fBmongerimport\fP utility, which
provides the inverse “importing” capability.
.sp
Run \fI\%mongerexport\fP from the system command line, not the \fBmonger\fP shell.
.SH CONSIDERATIONS
.sp
\fBWARNING:\fP
.INDENT 0.0
.INDENT 3.5
Avoid using \fBmongerimport\fP and \fI\%mongerexport\fP for
full instance production backups. They do not reliably preserve all rich
BSON data types, because JSON can only represent a subset
of the types supported by BSON. Use \fBmongerdump\fP
and \fBmongerrestore\fP as described in /core/backups for this
kind of functionality.
.UNINDENT
.UNINDENT
.sp
\fI\%mongerexport\fP must be run directly from the system command line.
.sp
To preserve type information, \fI\%mongerexport\fP and \fBmongerimport\fP
uses the strict mode representation
for certain types.
.sp
For example, the following insert operation in the \fBmonger\fP
shell uses the shell mode representation for the BSON types
\fBdata_date\fP and \fBdata_numberlong\fP:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
use test
db.traffic.insert( { _id: 1, volume: NumberLong(\(aq2980000\(aq), date: new Date() } )
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The argument to \fBdata_numberlong\fP must be quoted to avoid potential
loss of accuracy.
.sp
Use \fI\%mongerexport\fP to export the data:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
mongerexport \-\-db test \-\-collection traffic \-\-out traffic.json
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
The exported data is in strict mode representation to preserve type information:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
{ "_id" : 1, "volume" : { "$numberLong" : "2980000" }, "date" : { "$date" : "2014\-03\-13T13:47:42.483\-0400" } }
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
See /reference/mongerdb\-extended\-json for a complete list of
these types and the representations used.
.SH REQUIRED ACCESS
.sp
\fI\%mongerexport\fP requires read access on the target database.
.sp
Ensure that the connecting user possesses, at a minimum, the \fBread\fP
role on the target database.
.sp
When connecting to a \fBmongerd\fP or \fBmongers\fP that enforces
/core/authentication, ensure you use the required security
parameters based on the configured
authentication mechanism\&.
.SH READ PREFERENCE
.sp
\fI\%mongerexport\fP defaults to \fBprimary\fP read
preference when connected to a \fBmongers\fP
or a replica set\&.
.sp
You can override the default read preference using the
\fI\%\-\-readPreference\fP option.
.sp
\fBIMPORTANT:\fP
.INDENT 0.0
.INDENT 3.5
Using a non\-primary read preference on a \fBmongers\fP may
produce inconsistencies in data, including duplicates or missing
documents.
.UNINDENT
.UNINDENT
.SH OPTIONS
.sp
Changed in version 3.0.0: \fI\%mongerexport\fP removed the \fB\-\-dbpath\fP as well as related
\fB\-\-directoryperdb\fP and \fB\-\-journal\fP options. To use
\fI\%mongerexport\fP, you must run \fI\%mongerexport\fP against a running
\fBmongerd\fP or \fBmongers\fP instance as appropriate.

.sp
Changed in version 3.0.0: \fI\%mongerexport\fP removed the \fB\-\-csv\fP option. Use the
\fI\%\-\-type=csv\fP option to specify CSV format
for the output.

.INDENT 0.0
.TP
.B mongerexport
.UNINDENT
.INDENT 0.0
.TP
.B \-\-help
Returns information on the options and use of \fBmongerexport\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-verbose, \-v
Increases the amount of internal reporting returned on standard output
or in log files. Increase the verbosity with the \fB\-v\fP form by
including the option multiple times, (e.g. \fB\-vvvvv\fP\&.)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-quiet
Runs \fBmongerexport\fP in a quiet mode that attempts to limit the amount
of output.
.sp
This option suppresses:
.INDENT 7.0
.IP \(bu 2
output from database commands
.IP \(bu 2
replication activity
.IP \(bu 2
connection accepted events
.IP \(bu 2
connection closed events
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-version
Returns the \fBmongerexport\fP release number.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-uri <connectionString>
New in version 3.4.6.

.sp
Specify a resolvable URI
connection string for the \fBmongerd\fP to which to
connect.
.sp
The following is the standard
URI connection scheme:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
mongerdb://[username:password@]host1[:port1][,host2[:port2],...[,hostN[:portN]]][/[database][?options]]
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
For detailed explanations of the components of this string, refer to
the
Connection String URI Format
documentation.
.sp
\fBIMPORTANT:\fP
.INDENT 7.0
.INDENT 3.5
The following \fI\%mongerexport\fP options are incompatible with the
\fB\-\-uri\fP option. Instead, specify these options as part of your
\fB\-\-uri\fP connection string when applicable:
.INDENT 0.0
.IP \(bu 2
\fB\-\-host\fP
.IP \(bu 2
\fB\-\-port\fP
.IP \(bu 2
\fB\-\-db\fP
.IP \(bu 2
\fB\-\-username\fP
.IP \(bu 2
\fB\-\-password\fP (when specifying the password as part of the
URI connection string)
.IP \(bu 2
\fB\-\-authenticationDatabase\fP
.IP \(bu 2
\fB\-\-authenticationMechanism\fP
.UNINDENT
.UNINDENT
.UNINDENT
.UNINDENT
.INDENT 0.0
.TP
.B \-\-host <hostname><:port>, \-h <hostname><:port>
\fIDefault\fP: localhost:27017
.sp
Specifies a resolvable hostname for the \fBmongerd\fP to which to
connect. By default, the \fBmongerexport\fP attempts to connect to a MongerDB
instance running on the localhost on port number \fB27017\fP\&.
.sp
To connect to a replica set, specify the
\fBreplSetName\fP and a seed list of set members, as in
the following:
.INDENT 7.0
.INDENT 3.5
.sp
.nf
.ft C
<replSetName>/<hostname1><:port>,<hostname2><:port>,<...>
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
You can always connect directly to a single MongerDB instance by
specifying the host and port number directly.
.sp
Changed in version 3.0.0: If you use IPv6 and use the \fB<address>:<port>\fP format, you must
enclose the portion of an address and port combination in
brackets (e.g. \fB[<address>]\fP).

.UNINDENT
.INDENT 0.0
.TP
.B \-\-port <port>
\fIDefault\fP: 27017
.sp
Specifies the TCP port on which the MongerDB instance listens for
client connections.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ipv6
\fIRemoved in version 3.0.\fP
.sp
Enables IPv6 support and allows \fBmongerexport\fP to connect to the
MongerDB instance using an IPv6 network. Prior to MongerDB 3.0, you
had to specify \fI\%\-\-ipv6\fP to use IPv6. In MongerDB 3.0 and later, IPv6
is always enabled.
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ssl
New in version 2.6.

.sp
Enables connection to a \fBmongerd\fP or \fBmongers\fP that has
TLS/SSL support enabled.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslCAFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains the root certificate chain
from the Certificate Authority. Specify the file name of the
\fB\&.pem\fP file using relative or absolute paths.
.sp
Starting in version 3.4, if \fB\-\-sslCAFile\fP or \fBssl.CAFile\fP is not
specified and you are not using x.509 authentication, the
system\-wide CA certificate store will be used when connecting to an
TLS/SSL\-enabled server.
.sp
If using x.509 authentication, \fB\-\-sslCAFile\fP or \fBssl.CAFile\fP
must be specified.
.sp
\fBWARNING:\fP
.INDENT 7.0
.INDENT 3.5
\fBVersion 3.2 and earlier:\fP For TLS/SSL connections (\fB\-\-ssl\fP) to
\fBmongerd\fP and \fBmongers\fP, if the \fBmongerexport\fP runs without the
\fI\%\-\-sslCAFile\fP, \fBmongerexport\fP will not attempt
to validate the server certificates. This creates a vulnerability
to expired \fBmongerd\fP and \fBmongers\fP certificates as
well as to foreign processes posing as valid \fBmongerd\fP or
\fBmongers\fP instances. Ensure that you \fIalways\fP specify the
CA file to validate the server certificates in cases where
intrusion is a possibility.
.UNINDENT
.UNINDENT
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyFile <filename>
New in version 2.6.

.sp
Specifies the \fB\&.pem\fP file that contains both the TLS/SSL certificate
and key. Specify the file name of the \fB\&.pem\fP file using relative
or absolute paths.
.sp
This option is required when using the \fI\%\-\-ssl\fP option to connect
to a \fBmongerd\fP or \fBmongers\fP that has
\fBCAFile\fP enabled \fIwithout\fP
\fBallowConnectionsWithoutCertificates\fP\&.
.sp
For more information about TLS/SSL and MongerDB, see
/tutorial/configure\-ssl and
/tutorial/configure\-ssl\-clients .
.UNINDENT
.INDENT 0.0
.TP
.B \-\-sslPEMKeyPassword <value>
New in version 2.6.

.sp

debian/changelog
mongerdb (2.5.4) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/12898

 -- Ernie Hershey <ernie.hershey@mongerdb.com>  Thu, 19 Dec 2013 00:41:46 -0500

mongerdb (2.1.2) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10894

 -- Richard Kreuter <richard@10gen.com>  Tue, 26 Jun 2012 16:56:28 -0500

mongerdb (1.9.2) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10261

 -- Richard Kreuter <richard@10gen.com>  Thu, 11 Aug 2011 16:56:28 -0500

mongerdb (1.9.1) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10261

 -- Richard Kreuter <richard@10gen.com>  Tue, 26 Jul 2011 16:56:28 -0500

mongerdb (1.9.0) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10232

 -- Richard Kreuter <richard@10gen.com>  Sun, 01 May 2011 16:56:28 -0500

mongerdb (1.8.0-rc0) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10245

 -- Richard Kreuter <richard@10gen.com>  Tue, 22 Feb 2011 16:56:28 -0500

mongerdb (1.7.6) unstable; urgency=low

  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10223

 -- Richard Kreuter <richard@10gen.com>  Wed, 26 Jan 2011 16:56:28 -0500

mongerdb (1.7.5) unstable; urgency=low

  * durability journalingin place
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10214

 -- Richard Kreuter <richard@10gen.com>  Wed, 26 Jan 2011 16:56:28 -0500

mongerdb (1.7.4) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10210

 -- Richard Kreuter <richard@10gen.com>  Mon, 20 Dec 2010 16:56:28 -0500

mongerdb (1.7.3) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10195

 -- Richard Kreuter <richard@10gen.com>  Tue, 16 Nov 2010 16:56:28 -0500

mongerdb (1.7.2) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10194

 -- Richard Kreuter <richard@10gen.com>  Fri, 29 Oct 2010 16:56:28 -0500

mongerdb (1.7.1) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes
  * see http://jira.mongerdb.org/browse/SERVER/fixforversion/10193

 -- Richard Kreuter <richard@10gen.com>  Tue, 28 Sep 2010 16:56:28 -0500

mongerdb (1.7.0) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Fri, 03 Sep 2010 16:56:28 -0500

mongerdb (1.5.8) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Tue, 03 Aug 2010 16:56:28 -0500

mongerdb (1.5.7) unstable; urgency=low

  * sharding lots of changes
  * replica_sets lots of changes

 -- Richard Kreuter <richard@10gen.com>  Fri, 30 Jul 2010 16:56:28 -0500


mongerdb (1.5.6) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10179

 -- Richard Kreuter <richard@10gen.com>  Sat, 24 Jul 2010 16:56:28 -0500

mongerdb (1.5.5) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Fri, 16 Jul 2010 16:56:28 -0500

mongerdb (1.5.4) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Fri, 2 Jul 2010 16:56:28 -0500

mongerdb (1.5.3) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10157

 -- Richard Kreuter <richard@10gen.com>  Thu, 17 Jun 2010 16:56:28 -0500

mongerdb (1.5.2) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10143

 -- Richard Kreuter <richard@10gen.com>  Wed, 27 May 2010 16:56:28 -0500

mongerdb (1.5.1) unstable; urgency=low

  * sharding lots of changes, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10142

 -- Richard Kreuter <richard@10gen.com>  Wed, 3 May 2010 16:56:28 -0500

mongerdb (1.5.0) unstable; urgency=low

  * replication w & real-time, see http://jira.mongerdb.org/browse/SERVER/fixforversion/10125

 -- Richard Kreuter <richard@10gen.com>  Wed, 22 Mar 2010 16:56:28 -0500


mongerdb (1.3.5) unstable; urgency=low

  * bug fixes

 -- Richard Kreuter <richard@10gen.com>  Wed, 22 Mar 2010 16:56:28 -0500

mongerdb (1.3.4) unstable; urgency=low

  * bufg fixes

 -- Richard Kreuter <richard@10gen.com>  Wed, 17 Mar 2010 16:56:28 -0500

mongerdb (1.3.3) unstable; urgency=low

  * geo

 -- Richard Kreuter <richard@10gen.com>  Fri, 05 Feb 2010 16:56:28 -0500

mongerdb (1.3.2) unstable; urgency=low

  * munged debian files


src/mongo/gotools/src/github.com/mongodb/mongo-tools/mongoexport/mongoexport.go
// Copyright (C) MongoDB, Inc. 2014-present.
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may
// not use this file except in compliance with the License. You may obtain
// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0

// Package mongerexport produces a JSON or CSV export of data stored in a MongerDB instance.
package mongerexport

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/mongerdb/monger-tools-common/bsonutil"
	"github.com/mongerdb/monger-tools-common/db"
	"github.com/mongerdb/monger-tools-common/json"
	"github.com/mongerdb/monger-tools-common/log"
	"github.com/mongerdb/monger-tools-common/options"
	"github.com/mongerdb/monger-tools-common/progress"
	"github.com/mongerdb/monger-tools-common/util"
	"go.mongerdb.org/monger-driver/bson"
	"go.mongerdb.org/monger-driver/monger"
	driverOpts "go.mongerdb.org/monger-driver/monger/options"
	"go.mongerdb.org/monger-driver/monger/readpref"
)

// Output types supported by mongerexport.
const (
	CSV                            = "csv"
	JSON                           = "json"
	watchProgressorUpdateFrequency = 8000
)

// jsonFormat is the type for all valid extended JSON formats to output.
type jsonFormat string

const (
	canonical jsonFormat = "canonical"
	relaxed              = "relaxed"
)

const (
	progressBarLength   = 24
	progressBarWaitTime = time.Second
)

// MongerExport is a container for the user-specified options and
// internal state used for running mongerexport.
type MongerExport struct {
	// generic monger tool options
	ToolOptions *options.ToolOptions

	// OutputOpts controls options for how the exported data should be formatted
	OutputOpts *OutputFormatOptions

	InputOpts *InputOptions

	// for connecting to the db
	SessionProvider *db.SessionProvider
	ExportOutput    ExportOutput

	ProgressManager progress.Manager

	// Cached version of the collection info
	collInfo *db.CollectionInfo
}

// ExportOutput is an interface that specifies how a document should be formatted
// and written to an output stream.
type ExportOutput interface {
	// WriteHeader outputs any pre-record headers that are written once
	// per output file.
	WriteHeader() error

	// WriteRecord writes the given document to the given io.Writer according to
	// the format supported by the underlying ExportOutput implementation.
	ExportDocument(bson.D) error

	// WriteFooter outputs any post-record headers that are written once per
	// output file.
	WriteFooter() error

	// Flush writes any pending data to the underlying I/O stream.
	Flush() error
}

// New constructs a new MongerExport instance from the provided options.
func New(opts Options) (*MongerExport, error) {
	exporter := &MongerExport{
		ToolOptions: opts.ToolOptions,
		OutputOpts:  opts.OutputFormatOptions,
		InputOpts:   opts.InputOptions,
	}

	err := exporter.validateSettings()
	if err != nil {
		return nil, util.SetupError{
			Err:     err,
			Message: util.ShortUsage("mongerexport"),
		}
	}

	provider, err := db.NewSessionProvider(*opts.ToolOptions)
	if err != nil {
		return nil, util.SetupError{Err: err}
	}

	log.Logvf(log.Always, "connected to: %v", opts.URI.ConnectionString)

	isMongers, err := provider.IsMongers()
	if err != nil {
		provider.Close()
		return nil, util.SetupError{Err: err}
	}

	// warn if we are trying to export from a secondary in a sharded cluster
	pref := opts.ToolOptions.ReadPreference
	if isMongers && pref != nil && pref.Mode() != readpref.PrimaryMode {
		log.Logvf(log.Always, db.WarningNonPrimaryMongersConnection)
	}

	progressManager := progress.NewBarWriter(log.Writer(0), progressBarWaitTime, progressBarLength, false)
	progressManager.Start()

	exporter.SessionProvider = provider
	exporter.ProgressManager = progressManager
	return exporter, nil
}

// Close cleans up all the resources for a MongerExport instance.
func (exp *MongerExport) Close() {
	exp.SessionProvider.Close()
	if barWriter, ok := exp.ProgressManager.(*progress.BarWriter); ok {
		barWriter.Stop()
	}
}

// validateSettings returns an error if any settings specified on the command line
// were invalid, or nil if they are valid.
func (exp *MongerExport) validateSettings() error {
	// Namespace must have a valid database if none is specified,
	// use 'test'
	if exp.ToolOptions.Namespace.DB == "" {
		exp.ToolOptions.Namespace.DB = "test"
	}
	err := util.ValidateDBName(exp.ToolOptions.Namespace.DB)
	if err != nil {
		return err
	}

	if exp.ToolOptions.Namespace.Collection == "" {
		return fmt.Errorf("must specify a collection")
	}
	if err = util.ValidateCollectionGrammar(exp.ToolOptions.Namespace.Collection); err != nil {
		return err
	}

	exp.OutputOpts.Type = strings.ToLower(exp.OutputOpts.Type)

	if exp.OutputOpts.CSVOutputType {
		log.Logv(log.Always, "csv flag is deprecated; please use --type=csv instead")
		exp.OutputOpts.Type = CSV
	}

	if exp.OutputOpts.Type == "" {
		// special error for an empty type value
		return fmt.Errorf("--type cannot be empty")
	}
	if exp.OutputOpts.Type != CSV && exp.OutputOpts.Type != JSON {
		return fmt.Errorf("invalid output type '%v', choose 'json' or 'csv'", exp.OutputOpts.Type)
	}

	if exp.OutputOpts.JSONFormat != canonical && exp.OutputOpts.JSONFormat != relaxed {
		return fmt.Errorf("invalid JSON format '%v', choose 'relaxed' or 'canonical'", exp.OutputOpts.JSONFormat)
	}

	if exp.InputOpts.Query != "" && exp.InputOpts.ForceTableScan {
		return fmt.Errorf("cannot use --forceTableScan when specifying --query")
	}

	if exp.InputOpts.Query != "" && exp.InputOpts.QueryFile != "" {
		return fmt.Errorf("either --query or --queryFile can be specified as a query option")
	}

	if exp.InputOpts != nil && exp.InputOpts.HasQuery() {
		content, err := exp.InputOpts.GetQuery()
		if err != nil {
			return err
		}
		_, err2 := getObjectFromByteArg(content)
		if err2 != nil {
			return err2
		}
	}

	if exp.InputOpts != nil && exp.InputOpts.Sort != "" {
		_, err := getSortFromArg(exp.InputOpts.Sort)
		if err != nil {
			return err
		}
	}
	return nil
}

// GetOutputWriter opens and returns an io.WriteCloser for the output
// options or nil if none is set. The caller is responsible for closing it.
func (exp *MongerExport) GetOutputWriter() (io.WriteCloser, error) {
	if exp.OutputOpts.OutputFile != "" {
		// If the directory in which the output file is to be
		// written does not exist, create it
		fileDir := filepath.Dir(exp.OutputOpts.OutputFile)
		err := os.MkdirAll(fileDir, 0750)
		if err != nil {
			return nil, err
		}

		file, err := os.Create(util.ToUniversalPath(exp.OutputOpts.OutputFile))
		if err != nil {
			return nil, err
		}
		return file, err
	}
	// No writer, so caller should assume Stdout (or some other reasonable default)
	return nil, nil
}

// Take a comma-delimited set of field names and build a selector doc for query projection.
// For fields containing a dot '.', we project the entire top-level portion.
// e.g. "a,b,c.d.e,f.$" -> {a:1, b:1, "c":1, "f.$": 1}.
func makeFieldSelector(fields string) bson.M {
	selector := bson.M{"_id": 1}
	if fields == "" {
		return selector
	}

	for _, field := range strings.Split(fields, ",") {
		// Projections like "a.0" work fine for nested documents not for arrays
		// - if passed directly to mongerd. To handle this, we have to retrieve
		// the entire top-level document and then filter afterwards. An exception
		// is made for '$' projections - which are passed directly to mongerd.
		if i := strings.LastIndex(field, "."); i != -1 && field[i+1:] != "$" {
			field = field[:strings.Index(field, ".")]
		}
		selector[field] = 1
	}
	return selector
}

// getCount returns an estimate of how many documents the cursor will fetch
// It always returns Limit if there is a limit, assuming that in general
// limits will less then the total possible.
// If there is a query and no limit then it returns 0, because it's too expensive to count the query.
// If the collection is a view then it returns 0, because it is too expensive to count the view.
// Otherwise it returns the count minus the skip
func (exp *MongerExport) getCount() (int64, error) {
	session, err := exp.SessionProvider.GetSession()
	if err != nil {
		return 0, err
	}
	if exp.InputOpts != nil && exp.InputOpts.Limit != 0 {
		return exp.InputOpts.Limit, nil
	}
	if exp.InputOpts != nil && exp.InputOpts.Query != "" {
		return 0, nil
	}
	coll := session.Database(exp.ToolOptions.Namespace.DB).Collection(exp.ToolOptions.Namespace.Collection)

	if exp.collInfo.IsView() {
		return 0, nil
	}

	c, err := coll.EstimatedDocumentCount(nil)
	if err != nil {
		return 0, err
	}

	var skip int64
	if exp.InputOpts != nil {
		skip = exp.InputOpts.Skip
	}
	if skip > c {
		c = 0
	} else {
		c -= skip
	}
	return c, nil
}

// getCursor returns a cursor that can be iterated over to get all the documents
// to export, based on the options given to mongerexport. Also returns the
// associated session, so that it can be closed once the cursor is used up.
func (exp *MongerExport) getCursor() (*monger.Cursor, error) {
	findOpts := driverOpts.Find()

	if exp.InputOpts != nil && exp.InputOpts.Sort != "" {
		sortD, err := getSortFromArg(exp.InputOpts.Sort)
		if err != nil {
			return nil, err
		}

		findOpts.SetSort(sortD)
	}

	query := map[string]interface{}{}
	if exp.InputOpts != nil && exp.InputOpts.HasQuery() {
		var err error
		content, err := exp.InputOpts.GetQuery()
		if err != nil {
			return nil, err
		}
		query, err = getObjectFromByteArg(content)
		if err != nil {
			return nil, err
		}
	}

	session, err := exp.SessionProvider.GetSession()
	if err != nil {
		return nil, err
	}
	coll := session.Database(exp.ToolOptions.Namespace.DB).Collection(exp.ToolOptions.Namespace.Collection)

	// don't snapshot if we've been asked not to,
	// or if we cannot because  we are querying, sorting, or if the collection is a view
	if !exp.InputOpts.ForceTableScan && len(query) == 0 && exp.InputOpts != nil && exp.InputOpts.Sort == "" &&
		!exp.collInfo.IsView() && !exp.collInfo.IsSystemCollection() {

		// Use a hint on the _id index instead of the deprecated snapshot option
		findOpts.SetHint(bson.D{{"_id", 1}})
	}

	if exp.InputOpts != nil {
		findOpts.SetSkip(exp.InputOpts.Skip)
	}
	if exp.InputOpts != nil {
		findOpts.SetLimit(exp.InputOpts.Limit)
	}

	if len(exp.OutputOpts.Fields) > 0 {
		findOpts.SetProjection(makeFieldSelector(exp.OutputOpts.Fields))
	}

	return coll.Find(nil, query, findOpts)
}

// verifyCollectionExists checks if the collection exists. If it does, a copy of the collection info will be cached
// on the receiver. If the collection does not exist and AssertExists was specified, a non-nil error is returned.
func (exp *MongerExport) verifyCollectionExists() (bool, error) {
	session, err := exp.SessionProvider.GetSession()
	if err != nil {
		return false, err
	}

	coll := session.Database(exp.ToolOptions.Namespace.DB).Collection(exp.ToolOptions.Namespace.Collection)
	exp.collInfo, err = db.GetCollectionInfo(coll)
	if err != nil {
		return false, err
	}

	// If the collection doesn't exist, GetCollectionInfo will return nil
	if exp.collInfo == nil {
		var collInfoErr error
		if exp.InputOpts.AssertExists {
			collInfoErr = fmt.Errorf("collection '%s' does not exist", exp.ToolOptions.Namespace.Collection)
		}

		return false, collInfoErr
	}

	return true, nil
}

// Internal function that handles exporting to the given writer. Used primarily
// for testing, because it bypasses writing to the file system.
func (exp *MongerExport) exportInternal(out io.Writer) (int64, error) {
	// Check if the collection exists before starting export
	exists, err := exp.verifyCollectionExists()
	if err != nil || !exists {
		return 0, err
	}

	max, err := exp.getCount()
	if err != nil {
		return 0, err
	}

	watchProgressor := progress.NewCounter(int64(max))
	if exp.ProgressManager != nil {
		name := fmt.Sprintf("%v.%v", exp.ToolOptions.Namespace.DB, exp.ToolOptions.Namespace.Collection)
		exp.ProgressManager.Attach(name, watchProgressor)
		defer exp.ProgressManager.Detach(name)
	}

	exportOutput, err := exp.getExportOutput(out)
	if err != nil {
		return 0, err
	}

	cursor, err := exp.getCursor()
	if err != nil {
		return 0, err
	}
	defer cursor.Close(nil)

	// Write headers
	err = exportOutput.WriteHeader()
	if err != nil {
		return 0, err
	}

	docsCount := int64(0)

	// Write document content
	for cursor.Next(nil) {
		var result bson.D
		if err := cursor.Decode(&result); err != nil {
			return docsCount, err
		}

		err := exportOutput.ExportDocument(result)
		if err != nil {
			return docsCount, err
		}
		docsCount++
		if docsCount%watchProgressorUpdateFrequency == 0 {
			watchProgressor.Set(docsCount)
		}
	}
	watchProgressor.Set(docsCount)
	if err := cursor.Err(); err != nil {
		return docsCount, err
	}

	// Write footers
	err = exportOutput.WriteFooter()
	if err != nil {
		return docsCount, err
	}
	exportOutput.Flush()
	return docsCount, nil
}

// Export executes the entire export operation. It returns an integer of the count
// of documents successfully exported, and a non-nil error if something went wrong
// during the export operation.
func (exp *MongerExport) Export(out io.Writer) (int64, error) {
	count, err := exp.exportInternal(out)
	return count, err
}

// getExportOutput returns an implementation of ExportOutput which can handle
// transforming BSON documents into the appropriate output format and writing
// them to an output stream.
func (exp *MongerExport) getExportOutput(out io.Writer) (ExportOutput, error) {
	if exp.OutputOpts.Type == CSV {
		// TODO what if user specifies *both* --fields and --fieldFile?
		var fields []string
		var err error
		if len(exp.OutputOpts.Fields) > 0 {
			fields = strings.Split(exp.OutputOpts.Fields, ",")
		} else if exp.OutputOpts.FieldFile != "" {
			fields, err = util.GetFieldsFromFile(exp.OutputOpts.FieldFile)
			if err != nil {
				return nil, err
			}
		} else {
			return nil, fmt.Errorf("CSV mode requires a field list")
		}

		exportFields := make([]string, 0, len(fields))
		for _, field := range fields {
			// for '$' field projections, exclude '.$' from the field name
			if i := strings.LastIndex(field, "."); i != -1 && field[i+1:] == "$" {
				exportFields = append(exportFields, field[:i])
			} else {
				exportFields = append(exportFields, field)
			}
		}

		return NewCSVExportOutput(exportFields, exp.OutputOpts.NoHeaderLine, out), nil
	}
	return NewJSONExportOutput(exp.OutputOpts.JSONArray, exp.OutputOpts.Pretty, out, exp.OutputOpts.JSONFormat), nil
}

// getObjectFromByteArg takes an object in extended JSON, and converts it to an object that
// can be passed straight to db.collection.find(...) as a query or sort criteria.
// Returns an error if the string is not valid JSON, or extended JSON.
func getObjectFromByteArg(queryRaw []byte) (map[string]interface{}, error) {
	parsedJSON := map[string]interface{}{}
	err := json.Unmarshal(queryRaw, &parsedJSON)
	if err != nil {
		return nil, fmt.Errorf("query '%v' is not valid JSON: %v", queryRaw, err)
	}

	err = bsonutil.ConvertLegacyExtJSONDocumentToBSON(parsedJSON)
	if err != nil {
		return nil, err
	}
	return parsedJSON, nil
}

// getSortFromArg takes a sort specification in JSON and returns it as a bson.D
// object which preserves the ordering of the keys as they appear in the input.
func getSortFromArg(queryRaw string) (bson.D, error) {
	parsedJSON := bson.D{}
	err := json.Unmarshal([]byte(queryRaw), &parsedJSON)
	if err != nil {
		return nil, fmt.Errorf("query '%v' is not valid JSON: %v", queryRaw, err)
	}
	// TODO: verify sort specification before returning a nil error
	return parsedJSON, nil
}
